### PR TITLE
ControlledTree: Fix tree rendering with stale data

### PR DESCRIPTION
**Pull request by saskliutas**
## Changes

Fixed `useTreeModel` hook to make one more render when `TreeModelSource` changes to make sure that tree is rendered with latest data in case tree model was modified between `TreeModelSource` change and `onModelChanged` event listener was added.

Replaced `useRerender` with `useState` because `rerender` function had new reference on each render and caused `onModelChanged` event listener to be removed and added on each render.

## Testing

Added unit test.

### CODE REVIEW

**Comment by saskliutas** [comment]
@Mergifyio backport release/4.5.x

File: ui/components-react/src/test/tree/controlled/TreeHooks.test.tsx [hunk 1]
```diff
@@ -18,74 +18,125 @@ import {
...
-  beforeEach(() => {
-    modelSourceMock.reset();
-    onModelChangeEvent = new BeUiEvent<[TreeModel, TreeModelChanges]>();
-
-    modelSourceMock
-      .setup((x) => x.onModelChanged)
-      .returns(() => onModelChangeEvent);
-    modelSourceMock.setup((x) => x.getModel()).returns(() => testModel);
-  });
+  beforeEach(() => {});
```
Comment (grigasp) [inline ui/components-react/src/test/tree/controlled/TreeHooks.test.tsx:35]: seems unnecessary

File: ui/components-react/src/test/tree/controlled/TreeHooks.test.tsx [hunk 2]
```diff
@@ -18,74 +18,125 @@ import {
...
+
+    before(async () => {
+      await TestUtils.initializeUiComponents();
+    });
+
+    after(() => {
+      TestUtils.terminateUiComponents();
+    });
+
+    it("until subscription to `onModelChangedEvent`", async () => {
```
Comment (grigasp) [inline ui/components-react/src/test/tree/controlled/TreeHooks.test.tsx:129]: could you clarify the test name?

**Review by grigasp — COMMENTED** [review]
> Replaced `useRerender` with `useState` because `rerender` function had new reference on each render and caused `onModelChanged` event listener to be removed and added on each render.

Isn't that a bug in `useRerender`?

File: ui/components-react/src/test/tree/controlled/TreeHooks.test.tsx (hunk 2 above)
Comment (saskliutas) [reply ui/components-react/src/test/tree/controlled/TreeHooks.test.tsx]: Updated test name

**Comment by saskliutas** [comment]
> > Replaced `useRerender` with `useState` because `rerender` function had new reference on each render and caused `onModelChanged` event listener to be removed and added on each render.
> 
> Isn't that a bug in `useRerender`?

Not necessarily. If `useRerender` was added to be used in other places than `useEffect` there might be no need to memoize `rerender` function. Although I did a quick search and it looks like `useRerender` is not used anywhere anymore. So I will remove it.

**Review by GerardasB — APPROVED** [review]

**Review by grigasp — APPROVED** [review]

**Pull request merged**